### PR TITLE
Support adding columns in Delta Lake (ALTER TABLE ADD COLUMN)

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Support adding columns in Delta Lake

## Documentation

(x) No documentation is needed.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Delta Lake
* Add support for adding columns. ({issue}`12371`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java
Comment (homar) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:269]: Could you please add tests that will check:
- if spark can read table with added columns
- if optimize/vacuum works after adding columns ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (homar) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:903]: `columnMetadata` -> `newColumnMetadata` ?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:269]: > if spark can read table with added columns

Should I wait #11565?

Comment (homar) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: Verifying, even locally if that works would be great imho.
Regarding test - depends how long you would have to wait. 
@findinpath what is ETA for https://github.com/trinodb/trino/pull/11565 ?

Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: #11565 is in a good shape now. I am assuming that the PR will be ready to be merged in the following days.

Comment (homar) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: @ebyhr so my advice would be to wait for it and write a proper test

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:905]: Spark's Delta can cope with columns comments.

```
spark-sql> alter table t1 add columns (z string comment 'z comment');
Time taken: 1.54 seconds
spark-sql> desc t1;
x	int	
y	string	
z	string	z comment
		
# Partitioning		
Not partitioned		
Time taken: 0.075 seconds, Fetched 6 row(s)
```

Trino apparently does not interpret column comments.

```
trino> describe delta.default.t1;
 Column |  Type   | Extra | Comment 
--------+---------+-------+---------
 x      | integer |       |         
 y      | varchar |       |         
 z      | varchar |       |         
(3 rows)

Query 20220513_114236_00002_2rhtw, FINISHED, 1 node
Splits: 11 total, 11 done (100.00%)
0.38 [3 rows, 148B] [7 rows/s, 385B/s]
```

On the transaction log file the comment is to be found on the `metadata` field:

```
[root@hadoop-master /]# hdfs dfs -cat /user/hive/warehouse/t1/_delta_log/00000000000000000003.json
{"commitInfo":{"timestamp":1652442096774,"operation":"ADD COLUMNS","operationParameters":{"columns":"[{\"column\":{\"name\":\"z\",\"type\":\"string\",\"nullable\":true,\"metadata\":{\"comment\":\"z comment\"}}}]"},"readVersion":2,"isBlindAppend":true,"operationMetrics":{}}}
{"metaData":{"id":"7644697d-5e28-494a-8b6f-2f48e113eb7c","format":{"provider":"parquet","options":{}},"schemaString":"{\"type\":\"struct\",\"fields\":[{\"name\":\"x\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"y\",\"type\":\"string\",\"nullable\":true,\"metadata\":{}},{\"name\":\"z\",\"type\":\"string\",\"nullable\":true,\"metadata\":{\"comment\":\"z comment\"}}]}","partitionColumns":[],"configuration":{},"createdTime":1652441899688}}
```

**Review by alexjo2144 — CHANGES_REQUESTED** [review]
One important fix but the rest looks good to me

**Review by alexjo2144 — APPROVED** [review]

**Review by homar — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:905]: Thanks for providing the helpful information. Let me leave the comment support to #12424.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java
Comment (findepi) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: Does it mean you guys want to have this PR on hold until the other one lands?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:914]: The first entry will be silently overridden by another transaction as far as I confirmed. I mean, only one json log file will be generated after those two transactions.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: I could read new columns from Spark, but it prints below error at the firs access. There's no error from the 2nd access though. I'm little wondering if it needs fix or not. 
```sql
trino> ALTER TABLE delta.default.test ADD COLUMN x int;
ADD COLUMN

spark-sql> SELECT * FROM default.test;
ERROR DeltaLog: Change in the table id detected while updating snapshot.

1	NULL

spark-sql> SELECT * FROM default.test;
1	NULL
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:914]: The transaction log synchronizer should prevent a new json file from overwriting another

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:914]: Ah, I was testing with default query runner on local disk. Verified another transaction fails on S3.

Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:914]: > The transaction log synchronizer should prevent a new json file from overwriting another

Yes. 
But we're using `getMandatoryCurrentVersion` which checks, well, _current_ transaction, without checking whether some other thread did add another column with same name.

Am i missing something?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java
Comment (homar) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: I don't like this error but it looks like it returns the correct result anyway. It looks like spark cached something. What image did you use for testing? Can you try modyfing it like this https://github.com/trinodb/trino/commit/087868977cc7729621123bbc445384c9dfcae30f ?

Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: I tested on this product test PR #11565 

@findinpath may want to disable the cache in https://github.com/trinodb/trino/pull/11565/files#diff-8ecb8972788b7b6221c837540b00b020c2404ef64548ae563e02a9b1a66c8d14

Comment (homar) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: @ebyhr could you check that disabling the cache will actually help with this ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:914]: You're totally right. Instead of calling `getMandatoryCurrentVersion` we should be committing to `tableHandle.getReadVersion() + 1`

Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:914]: @alexjo2144 yeah, thanks for confirming.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java
Comment (findepi) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:277]: would running `REFRESH TABLE` in Spark help?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java]: Thanks for your help. Changed to `long commitVersion = handle.getReadVersion() + 1;`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:320]: @homar @findepi Disabling the cache didn't help. `REFRESH TABLE` logs the same error "ERROR DeltaLog: Change in the table id detected..." and subsequent `SELECT` don't show the error.

JFYI: The error comes from
https://github.com/delta-io/delta/blob/728bf902542077ce1c2e97ca67a53c53bb460c64/core/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala#L574-L575

Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:320]: It seems Spark reuse the same `metaData.id` when adding a new column.

```
trino> ALTER TABLE delta.default.test ADD COLUMN c4 int;
```

s3://presto-ci-test/test/_delta_log/00000000000000000004.json
```json
{"commitInfo":{"version":4,"timestamp":1653357796162,"userId":"yuya.ebihara","userName":"yuya.ebihara","operation":"ALTER TABLE","operationParameters":{"queryId":"20220524_020316_00004_imnpb"},"clusterId":"trino-dev-ffffffff-ffff-ffff-ffff-ffffffffffff","readVersion":0,"isolationLevel":"WriteSerializable","blindAppend":true}}
{"protocol":{"minReaderVersion":1,"minWriterVersion":2}}
{"metaData":{"id":"24b68017-cb79-4f0b-8ee7-e72a496bbaf4","format":{"provider":"parquet","options":{}},"schemaString":"{\"fields\":[{\"metadata\":{},\"name\":\"c1\",\"nullable\":true,\"type\":\"integer\"},{\"metadata\":{},\"name\":\"c2\",\"nullable\":true,\"type\":\"integer\"},{\"metadata\":{},\"name\":\"c3\",\"nullable\":true,\"type\":\"integer\"},{\"metadata\":{},\"name\":\"c4\",\"nullable\":true,\"type\":\"integer\"}],\"type\":\"struct\"}","partitionColumns":[],"configuration":{},"createdTime":1653357796162}}
```

```
spark-sql> ALTER TABLE default.test ADD COLUMN (x5 int);
```

s3://presto-ci-test/test/_delta_log/00000000000000000005.json
```json
{"commitInfo":{"timestamp":1653357938234,"operation":"ADD COLUMNS","operationParameters":{"columns":"[{\"column\":{\"name\":\"x5\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}}}]"},"readVersion":4,"isBlindAppend":true,"operationMetrics":{}}}
{"metaData":{"id":"24b68017-cb79-4f0b-8ee7-e72a496bbaf4","format":{"provider":"parquet","options":{}},"schemaString":"{\"type\":\"struct\",\"fields\":[{\"name\":\"c1\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"c2\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"c3\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"c4\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}},{\"name\":\"x5\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}}]}","partitionColumns":[],"configuration":{},"createdTime":1653357796162}}
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:226]: I will change to `ADD COLUMNS` for consistency with Spark.

**Comment by ebyhr** [comment]
@findepi @homar @findinpath @alexjo2144 Addressed comments.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:949]: tableId

**Review by findepi — APPROVED** [review]

**Review by findepi — APPROVED** [review]

**Comment by mosabua** [comment]
I guess rename and drop column are coming later..

**Comment by ebyhr** [comment]
@mosabua I agree with updating the docs.

**Comment by mosabua** [comment]
Maybe that means "ADD COLUMN" was the last one missing..

**Comment by Dearkano** [comment]
@mosabua I think RENAME COLUMN is not supported yet...at least in 370, when I rename the column, the query from spark is successful, but Trino returns NULL on that new column.

**Comment by mosabua** [comment]
we need to know where we are at with the current 381 release and @ebyhr can confirm in the code for us

**Comment by ebyhr** [comment]
> Trino returns NULL on that new column

@Dearkano It looks a bug in Delta Lake connector. Could you file an issue to https://github.com/trinodb/trino/issues/new?

**Pull request merged**
